File: packages/client/src/services/filesService.ts
import { BaseService } from './baseService.js';
import type { FileOptions } from '../types.js';

export interface BucketUploadOptions {
  /** Optional determinate fileId (for migration from Appwrite/Firebase) */
  fileId?: string;
}

export interface BucketUploadResult {
  fileId: string;
  filename: string;
  contentType: string;
  size: number;
  url: string;
}

export class FilesService extends BaseService {
  /**
   * URL untuk file yang ter-attach ke record (field type file)
   */
  getUrl(
    collection: string,
    recordId: string,
    filename: string,
    options: FileOptions = {}
  ): string {
    const base = this.client.baseUrl.replace(/\/$/, '');
    const pid = encodeURIComponent(this.client.projectId);
    const col = encodeURIComponent(collection);
    const rid = encodeURIComponent(recordId);
    const file = encodeURIComponent(filename);

    let url = `${base}/api/files/${pid}/${col}/${rid}/${file}`;

    if (options.thumb) {
      url += `?thumb=${encodeURIComponent(options.thumb)}`;
    }

    return url;
  }

  // ─── M35: Bucket-style storage (decoupled from records) ─────────────────────

  /**
   * URL untuk bucket file (decoupled, ala Appwrite getFileView).
   *
   * @param fileId ID dari bucketUpload
   */
  getBucketUrl(fileId: string): string {
    const base = this.client.baseUrl.replace(/\/$/, '');
    const pid = encodeURIComponent(this.client.projectId);
    return `${base}/api/files/${pid}/bucket/${encodeURIComponent(fileId)}`;
  }

  /**
   * Upload file ke bucket → return fileId + URL.
   * Decoupled dari records — file punya ID sendiri, bisa dipakai di mana saja.
   *
   * @param file File/Blob dari input element
   * @param filename Nama file
   * @param options Optional fileId untuk determinate ID (migration)
   */
  async uploadToBucket(
    file: File | Blob,
    filename?: string,
    options: BucketUploadOptions = {}
  ): Promise<BucketUploadResult> {
    const name = filename ?? (file instanceof File ? file.name : 'file');
    const formData = new FormData();
    formData.append('file', file, name);
    if (options.fileId) {
      formData.append('fileId', options.fileId);
    }

    const path = `api/p/${this.client.projectId}/storage/upload`;
    const result = await this.request<BucketUploadResult>(path, {
      method: 'POST',
      body: formData,
    });
    return result;
  }

  /**
   * Hapus file dari bucket.
   * User biasa hanya bisa hapus file miliknya sendiri.
   */
  async deleteBucketFile(fileId: string): Promise<boolean> {
    const path = `api/p/${this.client.projectId}/storage/bucket/${encodeURIComponent(fileId)}`;
    const result = await this.request<{ ok: boolean }>(path, { method: 'DELETE' });
    return result.ok;
  }

  /**
   * List bucket files milik user yang sedang login.
   */
  async listBucketFiles(): Promise<Array<{
    fileId: string;
    filename: string;
    contentType: string;
    size: number;
    uploadedAt: string;
    uploadedBy: string | null;
  }>> {
    const path = `api/p/${this.client.projectId}/storage/bucket`;
    const result = await this.request<{ files: Array<{
      fileId: string;
      filename: string;
      contentType: string;
      size: number;
      uploadedAt: string;
      uploadedBy: string | null;
    }> }>(path);
    return result.files;
  }
}

File: packages/kotlin-client/src/main/kotlin/id/wekanz/baseforge/files/FilesService.kt
package id.wekanz.baseforge.files

import id.wekanz.baseforge.internal.HttpCore
import id.wekanz.baseforge.internal.BaseForgeJson
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody

/** Hasil upload ke bucket — identik dengan `BucketUploadResult` di SDK TS. */
@Serializable
public data class BucketUploadResult(
    val fileId: String,
    val filename: String,
    val contentType: String,
    val size: Long,
    val url: String,
)

/** Satu entri dari [FilesService.listBucketFiles]. */
@Serializable
public data class BucketFileInfo(
    val fileId: String,
    val filename: String,
    val contentType: String,
    val size: Long,
    val uploadedAt: String,
    val uploadedBy: String? = null,
)

@Serializable
private data class ListBucketResponse(val files: List<BucketFileInfo> = emptyList())

@Serializable
private data class DeleteResponse(val ok: Boolean = false)

/**
 * Operasi file (bucket storage) — paritas dengan `filesService.ts` di SDK TS.
 *
 * Didapat lewat [id.wekanz.baseforge.BaseForge.files], jangan di-new langsung.
 *
 * Dua jenis URL berbeda yang penting dibedakan:
 * - [getUrl] — file yang ter-attach ke sebuah RECORD (field type file).
 * - [getBucketUrl] — file di BUCKET (decoupled dari record, punya fileId sendiri).
 *   Inilah yang dipakai dashboard untuk lampiran.
 */
public class FilesService internal constructor(
    private val http: HttpCore,
    private val baseUrl: String,
    private val projectId: String,
) {
    /**
     * URL untuk file yang ter-attach ke record (field type file).
     * Hanya MENYUSUN string — tidak melakukan request; klien (Coil/Glide) yang memuatnya.
     */
    public fun getUrl(
        collection: String,
        recordId: String,
        filename: String,
        thumb: String? = null,
    ): String {
        val base = baseUrl.trimEnd('/')
        var url = "$base/api/files/$projectId/$collection/$recordId/$filename"
        if (thumb != null) url += "?thumb=" + java.net.URLEncoder.encode(thumb, "UTF-8")
        return url
    }

    /** URL untuk bucket file (decoupled, ala Appwrite getFileView). */
    public fun getBucketUrl(fileId: String): String {
        val base = baseUrl.trimEnd('/')
        return "$base/api/files/$projectId/bucket/$fileId"
    }

    /**
     * Upload file ke bucket → return fileId + URL.
     *
     * [bytes] adalah isi file (mis. dari Bitmap.compress ke ByteArrayOutputStream di
     * Android). [fileId] opsional untuk determinate ID (migrasi dari Appwrite/Firebase).
     */
    public suspend fun uploadToBucket(
        bytes: ByteArray,
        filename: String,
        contentType: String = "application/octet-stream",
        fileId: String? = null,
    ): BucketUploadResult {
        val bodyBuilder = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("file", filename, bytes.toRequestBody(contentType.toMediaType()))
        if (fileId != null) bodyBuilder.addFormDataPart("fileId", fileId)

        val text = http.requestRaw(
            "POST",
            "api/p/$projectId/storage/upload",
            rawBody = bodyBuilder.build(),
        ) ?: error("empty upload response")
        return BaseForgeJson.decodeFromString(BucketUploadResult.serializer(), text)
    }

    /** Hapus file dari bucket. User biasa hanya bisa hapus file miliknya sendiri. */
    public suspend fun deleteBucketFile(fileId: String): Boolean {
        val text = http.request("DELETE", "api/p/$projectId/storage/bucket/$fileId") ?: return true
        return runCatching {
            BaseForgeJson.decodeFromString(DeleteResponse.serializer(), text).ok
        }.getOrDefault(true)
    }

    /** List bucket files milik user yang sedang login. */
    public suspend fun listBucketFiles(): List<BucketFileInfo> {
        val text = http.request("GET", "api/p/$projectId/storage/bucket") ?: "{}"
        return BaseForgeJson.decodeFromString(ListBucketResponse.serializer(), text).files
    }
}
